### PR TITLE
Make Function Env immutable

### LINKED ISSUE #1612: Calling a host function with environment in parallel results in data race
### Describe the bug

Causing a host function to be called in parallel leads to aliased mutable references to the function's `Env`. This is a data race and is undefined behavior.

Here's a complete example:

```rust
use wasmer::{Cranelift, Function, Instance, Module, Store, Value, JIT};

fn main() {
    let mut compiler_config = Cranelift::new();
    let engine = JIT::new(&mut compiler_config).engine();
    let store = Store::new(&engine);

    let wasm_source = "
    (module
        (type (;0;) (func (param i32) (result i32)))
        (import \"env\" \"host_function\" (func $host_function (type 0)))
        (func $guest_function (type 0) (param i32) (result i32)
          local.get 0
          call $host_function)
        (table (;0;) 1 1 funcref)
        (memory (;0;) 16)
        (global (;0;) (mut i32) (i32.const 1048576))
        (global (;1;) i32 (i32.const 1048576))
        (global (;2;) i32 (i32.const 1048576))
        (export \"memory\" (memory 0))
        (export \"guest_function\" (func $guest_function))
        (export \"__data_end\" (global 1))
        (export \"__heap_base\" (global 2)))
    ";
    let module = Module::new(&store, wasm_source).unwrap();
    let import_object = wasmer::imports! {
        "env" => {
            // host_function has a `u32` environment
            "host_function" => Function::new_native_with_env(&store, 10u32, host_function),
        }
    };
    let instance = Instance::new(&module, &import_object).unwrap();

    // spawn 8 threads to call guest_function in parallel
    let mut threads = Vec::new();
    for _ in 0..8 {
        let instance = instance.clone();
        threads.push(std::thread::spawn(move || {
            for _ in 0..1_000_000 {
                let guest_function = instance.exports.get_function("guest_function").unwrap();
                let _result = guest_function.call(&[Value::I32(1000)]).unwrap();
            }
        }));
    }

    for thread in threads {
        thread.join().unwrap();
    }
}

fn host_function(env: &mut u32, x: u32) -> u32 {
    *env += 1;
    if *env % 1_000_000 == 0 {
        println!("{}", env);
    }
    x
}

```

Running with `wasmer-1.0.0-alpha02.0` and `cargo run --release` gives me this output on Linux:

```
1000079
1000131
1000262
1001312
2000104
2000220
3000017
3000307
````

There's a pretty clear data race here—the correct output would be 

```
1000000
2000000
3000000
4000000
5000000
6000000
7000000
8000000
```

### Steps to reproduce
Run the above example with `cargo run --release`

### Expected behavior
This looks like the API for host function environments needs to be changed so the function doesn't receive a mutable reference to its `Env`.

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Alternative / related to #1625

Resolves #1612 

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by Hywan** [comment]
I agree `Env` must be immutable, and let the user use interior mutability (with `Cell`-like types) if necessary. Thanks for addressing that!

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

File: lib/api/Cargo.toml [hunk 1]
```diff
@@ -67,6 +67,8 @@ llvm = [
     "wasmer-compiler-llvm",
     "compiler",
 ]
+# enables internal features used by the deprecated API.
+deprecated = []
```
Comment (nlewycky) [inline lib/api/Cargo.toml:71]: Optional: really wish we didn't have to do this. Deprecated ought to be built on top of the API. Or put another way, anything deprecated does should be supported by our real API as something that other library consumer may also want to do.

Not that I'm offering an alternative. IIUC we considered wrapping it in a Mutex and rejected that?

File: lib/api/src/externals/function.rs [hunk 2]
```diff
@@ -235,7 +238,50 @@ impl Function {
...
+                vmctx,
+                signature,
+                call_trampoline: None,
+            },
+        }
+    }
+
+    /// Function used by the deprecated API to call a function with a `&mut` Env.
+    ///
+    /// This is  not a stable API and may be broken at any time.
```
Comment (nlewycky) [inline lib/api/src/externals/function.rs:261]:
```suggestion
    /// This is not a stable API and may be broken at any time.
```

File: lib/api/src/externals/function.rs [hunk 3]
```diff
@@ -989,6 +1043,18 @@ mod inner {
 
     impl HostFunctionKind for WithEnv {}
 
+    /// An empty struct to help Rust typing to determine
+    /// when a `HostFunction` does have an environment.
```
Comment (nlewycky) [inline lib/api/src/externals/function.rs:1047]:
```suggestion
    /// when a `HostFunction` has an environment.
```

File: lib/api/src/externals/function.rs [hunk 4]
```diff
@@ -989,6 +1043,18 @@ mod inner {
...
+    ///
+    /// This environment is passed by `&mut` and exists soley for the deprecated
```
Comment (nlewycky) [inline lib/api/src/externals/function.rs:1049]:
```suggestion
    /// This environment is passed by `&mut` and exists solely for the deprecated
```

File: lib/deprecated/runtime-core/src/vm.rs [hunk 5]
```diff
@@ -38,6 +40,9 @@ pub struct Ctx {
     pub data_finalizer: Option<fn(data: *mut c_void)>,
 }
 
+/// We mark `Ctx` as a legacy env that can be passed by `&mut`.
+unsafe impl LegacyEnv for Ctx {}
```
Comment (nlewycky) [inline lib/deprecated/runtime-core/src/vm.rs:44]: Optional: what about `UnsafeMutableEnv`?

File: lib/vm/src/instance.rs [hunk 6]
```diff
@@ -278,8 +278,8 @@ impl Instance {
     }
 
     /// Return a raw pointer to the vmctx used by compiled wasm code.
-    pub fn vmctx_ptr(&self) -> *const VMContext {
-        self.vmctx() as *const VMContext
+    pub fn vmctx_ptr(&self) -> *mut VMContext {
```
Comment (nlewycky) [inline lib/vm/src/instance.rs]: Where's the code that needs this to be mutable?

File: lib/api/Cargo.toml (hunk 1 above)
Comment (MarkMcCaskey) [reply lib/api/Cargo.toml:71]: > Not that I'm offering an alternative. IIUC we considered wrapping it in a Mutex and rejected that?

That was a different PR, if it were as simple as just conditionally putting something behind a mutex, then I agree.  I tried 2-3 times to make it work without editing the API and this is the only way I've been able to make it work.

I agree that it kind of sucks, but I figure the deprecated API won't be around forever so 🤷

File: lib/vm/src/instance.rs (hunk 6 above)
Comment (MarkMcCaskey) [reply lib/vm/src/instance.rs]: It's like that on master

https://github.com/wasmerio/wasmer/blob/master/lib/vm/src/instance.rs#L281-L283

I probably looked at master and got confused or something 🤷

Comment (MarkMcCaskey) [reply lib/vm/src/instance.rs]: Actually I don't even know how you commented on that, I don't see that in the diff 😆

File: lib/deprecated/runtime-core/src/vm.rs (hunk 5 above)
Comment (MarkMcCaskey) [reply lib/deprecated/runtime-core/src/vm.rs:44]: Yeah, I'm open to a new name, that does sound more descriptive

File: lib/c-api/wasmer_wasm.h [hunk 7]
```diff
@@ -29,9 +29,6 @@
 #  define DEPRECATED(message) __declspec(deprecated(message))
 #endif
 
-// The `jit` feature has been enabled for this build.
-#define WASMER_JIT_ENABLED
-
```
Comment (nlewycky) [inline lib/c-api/wasmer_wasm.h:34]: ??

**Review by nlewycky — APPROVED** [review]
IMO given how all the parts deprecated needs are now `unsafe` and documented as non-stable, you could remove the `deprecated` feature. A pro of this is that we don't have two possible builds of lib/api, with deprecated enabled and disabled. Maybe that doesn't matter.

File: lib/c-api/wasmer_wasm.h (hunk 7 above)
Comment (MarkMcCaskey) [reply lib/c-api/wasmer_wasm.h:34]: Oops, not sure how that ended up in there

**Review by syrusakbary — APPROVED** [review]
Looks great to merge, once the "deprecated" comment from Nick is addressed :)

**Pull request merged**